File: trust_primitive_engine/python/primitives/time_window.py
"""Deterministic inclusive time-window trust primitive."""

from __future__ import annotations

import re
from typing import Any

from engine import EvaluationState, Primitive, PrimitiveResult


EXPECTED_MEMBERS = {
    "requirement_id",
    "type",
    "not_before",
    "not_after",
}

IDENTIFIER_RE = re.compile(
    r"^[a-z0-9][a-z0-9._:/-]{1,127}[a-z0-9]$"
)

MAX_SAFE_INTEGER = 9_007_199_254_740_991


class TimeWindowPrimitive(Primitive):
    """Require evaluation_time to fall inside an inclusive interval."""

    TYPE = "time_window"

    def validate(self, value: dict[str, Any]) -> dict[str, Any]:
        unknown = sorted(set(value) - EXPECTED_MEMBERS)
        missing = sorted(EXPECTED_MEMBERS - set(value))

        if unknown:
            raise ValueError(f"time_window unknown members: {unknown}")
        if missing:
            raise ValueError(f"time_window missing members: {missing}")

        requirement_id = value["requirement_id"]
        not_before = value["not_before"]
        not_after = value["not_after"]

        if (
            not isinstance(requirement_id, str)
            or not IDENTIFIER_RE.fullmatch(requirement_id)
        ):
            raise ValueError(
                "requirements[].requirement_id is not a valid identifier"
            )

        for field, bound in (
            ("not_before", not_before),
            ("not_after", not_after),
        ):
            if (
                not isinstance(bound, int)
                or isinstance(bound, bool)
                or bound < 0
                or bound > MAX_SAFE_INTEGER
            ):
                raise ValueError(
                    f"requirements[].{field} must be an integer from "
                    f"0 to {MAX_SAFE_INTEGER}"
                )

        if not_before > not_after:
            raise ValueError(
                "time_window.not_before must be less than or equal to "
                "time_window.not_after"
            )

        return {
            "requirement_id": requirement_id,
            "type": self.TYPE,
            "not_before": not_before,
            "not_after": not_after,
        }

    def evaluate(
        self,
        requirement: dict[str, Any],
        state: EvaluationState,
    ) -> PrimitiveResult:
        not_before = requirement["not_before"]
        not_after = requirement["not_after"]
        evaluation_time = state.evaluation_time

        if evaluation_time is None:
            position = "missing"
        elif evaluation_time < not_before:
            position = "before"
        elif evaluation_time > not_after:
            position = "after"
        else:
            position = "inside"

        observed = {
            "evaluation_time": evaluation_time,
            "position": position,
        }
        expected = {
            "not_before": not_before,
            "not_after": not_after,
        }

        if position == "inside":
            return PrimitiveResult.satisfied_result(
                requirement_id=requirement["requirement_id"],
                primitive_type=self.TYPE,
                matched_signers=[],
                observed=observed,
                expected=expected,
            )

        return PrimitiveResult.unsatisfied_result(
            requirement_id=requirement["requirement_id"],
            primitive_type=self.TYPE,
            matched_signers=[],
            observed=observed,
            expected=expected,
            failure_code="TIME_WINDOW_NOT_SATISFIED",
        )

File: trust_primitive_engine/go/internal/primitives/temporal/time_window.go
package temporal

import (
	"agpprotocol.org/agp/trust-primitive-engine/internal/model"
	"agpprotocol.org/agp/trust-primitive-engine/internal/parser"
	"agpprotocol.org/agp/trust-primitive-engine/internal/primitives/contextvalue"
	"fmt"
)

const (
	TypeTimeWindow = "time_window"
	FailureCode    = "TIME_WINDOW_NOT_SATISFIED"
)

func Evaluate(requirement map[string]any, ctx model.Context) (map[string]any, bool, error) {
	requirementID, err := parser.AsString(requirement["requirement_id"], "requirement_id")
	if err != nil {
		return nil, false, err
	}
	notBefore, err := contextvalue.SafeInteger(requirement["not_before"], "not_before")
	if err != nil {
		return nil, false, err
	}
	notAfter, err := contextvalue.SafeInteger(requirement["not_after"], "not_after")
	if err != nil {
		return nil, false, err
	}
	if notBefore < 0 {
		return nil, false, fmt.Errorf(
			"time_window.not_before must be non-negative",
		)
	}
	if notAfter < 0 {
		return nil, false, fmt.Errorf(
			"time_window.not_after must be non-negative",
		)
	}
	if err := ValidateBounds(notBefore, notAfter); err != nil {
		return nil, false, err
	}

	position := "missing"
	var observedTime any
	if ctx.EvaluationTime != nil {
		observedTime = *ctx.EvaluationTime
		switch {
		case *ctx.EvaluationTime < notBefore:
			position = "before"
		case *ctx.EvaluationTime > notAfter:
			position = "after"
		default:
			position = "inside"
		}
	}

	satisfied := position == "inside"
	status := "unsatisfied"
	if satisfied {
		status = "satisfied"
	}

	result := map[string]any{
		"requirement_id":  requirementID,
		"type":            TypeTimeWindow,
		"status":          status,
		"matched_signers": []string{},
		"observed": map[string]any{
			"evaluation_time": observedTime,
			"position":        position,
		},
		"expected": map[string]any{
			"not_before": notBefore,
			"not_after":  notAfter,
		},
		"failure_code": nil,
	}
	if !satisfied {
		result["failure_code"] = FailureCode
	}
	return result, satisfied, nil
}

// ValidateBounds enforces the canonical non-negative inclusive interval.
func ValidateBounds(notBefore, notAfter int64) error {
	if notBefore < 0 {
		return fmt.Errorf(
			"time_window.not_before must be non-negative",
		)
	}
	if notAfter < 0 {
		return fmt.Errorf(
			"time_window.not_after must be non-negative",
		)
	}
	if notBefore > notAfter {
		return fmt.Errorf(
			"time_window.not_before must be less than or equal to " +
				"time_window.not_after",
		)
	}
	return nil
}
